src/TimeTrace.zig

```diff
@@ -0,0 +1,134 @@
+const std = @import("std");
+const InternPool = @import("InternPool.zig");
+const Zcu = @import("Zcu.zig");
+const TimeTrace = @This();
+
+gpa: std.mem.Allocator,
+enabled: bool = false,
+start_time: std.time.Instant,
+events: std.ArrayListUnmanaged(Event) = .{},
+mutex: std.Thread.Mutex = .{},
+
+pub const EventDesc = union(enum) {
+    // Sema events:
+    fn_body: InternPool.Index,
+    decl: Zcu.Decl.Index,
+
+    // Codegen events:
+    codegen_func: InternPool.Index,
+    codegen_decl: Zcu.Decl.Index,
+
+    end: void,
+};
+
+const Event = struct {
+    time: std.time.Instant,
+    pt: Zcu.PerThread,
+    desc: EventDesc,
```

Comment:
you could reduce overhead by reducing the size of this struct.
* 99% of the events are going to have the same Zcu
* you don't need 128 bits of time data per event
* EventDesc wastes 3 bytes per event due to padding

I guess it's fine though since this code is dead unless you ask for it.